dlt/destinations/synapse/sql_client.py

```diff
@@ -0,0 +1,222 @@
+import platform
+import struct
+from datetime import datetime, timedelta, timezone  # noqa: I251
+
+from dlt.common.destination import DestinationCapabilitiesContext
+
+import pyodbc
+
+from contextlib import contextmanager
+from typing import Any, AnyStr, ClassVar, Iterator, Optional, Sequence
+
+from dlt.destinations.exceptions import DatabaseTerminalException, DatabaseTransientException, DatabaseUndefinedRelation
+from dlt.destinations.typing import DBApi, DBApiCursor, DBTransaction
+from dlt.destinations.sql_client import DBApiCursorImpl, SqlClientBase, raise_database_error, raise_open_connection_error
+
+from dlt.destinations.synapse.configuration import SynapseCredentials
+from dlt.destinations.synapse import capabilities
+
+
+def handle_datetimeoffset(dto_value: bytes) -> datetime:
+    # ref: https://github.com/mkleehammer/pyodbc/issues/134#issuecomment-281739794
+    tup = struct.unpack("<6hI2h", dto_value)  # e.g., (2017, 3, 16, 10, 35, 18, 500000000, -6, 0)
+    return datetime(
+        tup[0], tup[1], tup[2], tup[3], tup[4], tup[5], tup[6] // 1000, timezone(timedelta(hours=tup[7], minutes=tup[8]))
+    )
+
+
+class PyOdbcSynapseClient(SqlClientBase[pyodbc.Connection], DBTransaction):
+
+    dbapi: ClassVar[DBApi] = pyodbc
+    capabilities: ClassVar[DestinationCapabilitiesContext] = capabilities()
+
+    def __init__(self, dataset_name: str, credentials: SynapseCredentials) -> None:
+        super().__init__(credentials.database, dataset_name)
+        self._conn: pyodbc.Connection = None
+        self.credentials = credentials
+
+    def open_connection(self) -> pyodbc.Connection:
+        self._conn = pyodbc.connect(
+            self.credentials.to_odbc_dsn(),
+            timeout=self.credentials.connect_timeout,
+        )
+        self._conn.add_output_converter(-155, handle_datetimeoffset)
+        self._conn.autocommit = True
+        print("Conn string: ", self.credentials.to_odbc_dsn())
+        return self._conn
+
+    @raise_open_connection_error
+    def close_connection(self) -> None:
+        if self._conn:
+            self._conn.close()
+            self._conn = None
+
+    @contextmanager
+    def begin_transaction(self) -> Iterator[DBTransaction]:
+        try:
+            self._conn.autocommit = False
+            yield self
+            self.commit_transaction()
+        except Exception:
+            self.rollback_transaction()
+            raise
+
+    @raise_database_error
+    def commit_transaction(self) -> None:
+        self._conn.commit()
+        self._conn.autocommit = True
+
+    @raise_database_error
+    def rollback_transaction(self) -> None:
+        self._conn.rollback()
+        self._conn.autocommit = True
+
+    @property
+    def native_connection(self) -> pyodbc.Connection:
+        return self._conn
+
+    def drop_dataset(self) -> None:
```

Comment:
why do you need to check if each table exists? can't you just drop all tables in the dataset like it is done in the mssql client?